File: python/bindings/MasterVector.cxx
// Optima is a C++ library for solving linear and non-linear constrained optimization problems.
//
// Copyright © 2020-2024 Allan Leal
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

// pybind11 includes
#include "pybind11.hxx"

// Optima includes
#include <Optima/MasterVector.hpp>
using namespace Optima;

void exportMasterVector(py::module& m)
{
    auto __add__      = [](const MasterVector& l, const MasterVector& r) -> MasterVector { return l + r; };
    auto __sub__      = [](const MasterVector& l, const MasterVector& r) -> MasterVector { return l - r; };
    auto __mul__      = [](const MasterVector& l, double r) -> MasterVector { return l * r; };
    auto __truediv__  = [](const MasterVector& l, double r) -> MasterVector { return l / r; };

    py::class_<MasterVector>(m, "MasterVector")
        .def(py::init<>())
        .def(py::init<const MasterDims&>())
        .def(py::init<Index, Index, Index>())
        .def(py::init<VectorView, Index, Index, Index>())
        .def(py::init<const MasterVector&>())
        .def(py::init<const MasterVectorRef&>())
        .def(py::init<const MasterVectorView&>())
        .def_property("x", [](MasterVector& s) -> VectorRef { return s.x; }, [](MasterVector& s, VectorView x) { s.x = x; })
        .def_property("p", [](MasterVector& s) -> VectorRef { return s.p; }, [](MasterVector& s, VectorView p) { s.p = p; })
        .def_property("w", [](MasterVector& s) -> VectorRef { return s.w; }, [](MasterVector& s, VectorView w) { s.w = w; })
        .def(py::self += py::self)
        .def(py::self -= py::self)
        .def(py::self *= double())
        .def(py::self /= double())
        .def("__add__", __add__)
        .def("__sub__", __sub__)
        .def("__mul__", __mul__)
        .def("__rmul__", __mul__)
        .def("__truediv__", __truediv__)
        .def("resize", &MasterVector::resize)
        .def("dot", [](const MasterVector& l, const MasterVector& r) { return l.dot(r); })
        .def("norm", &MasterVector::norm)
        .def("squaredNorm", &MasterVector::squaredNorm)
        .def("size", &MasterVector::size)
        .def("array", [](const MasterVector& self) { return Vector(self); })
        .def(double() * py::self)
        ;

    py::class_<MasterVectorRef>(m, "MasterVectorRef")
        .def(py::init<VectorRef, Index, Index, Index>())
        .def(py::init<MasterVector&>())
        .def(py::init<MasterVectorRef&>())
        .def_readwrite("x", &MasterVectorRef::x)
        .def_readwrite("p", &MasterVectorRef::p)
        .def_readwrite("w", &MasterVectorRef::w)
        .def("size", &MasterVectorRef::size)
        .def("array", [](const MasterVectorRef& self) { return Vector(self); })
        ;

    py::class_<MasterVectorView>(m, "MasterVectorView")
        .def(py::init<VectorView, Index, Index, Index>())
        .def(py::init<const MasterVector&>())
        .def(py::init<const MasterVectorRef&>())
        .def_readonly("x", &MasterVectorView::x)
        .def_readonly("p", &MasterVectorView::p)
        .def_readonly("w", &MasterVectorView::w)
        .def("size", &MasterVectorView::size)
        .def("array", [](const MasterVectorView& self) { return Vector(self); })
        ;

    py::implicitly_convertible<MasterVector, MasterVectorRef>();
    py::implicitly_convertible<MasterVector, MasterVectorView>();
    py::implicitly_convertible<MasterVectorRef, MasterVectorView>();
}

File: tests/MasterVector.py
# Optima is a C++ library for numerical solution of linear and nonlinear programing problems.
#
# Copyright © 2020-2024 Allan Leal
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.


from testing.optima import *
from testing.utils.matrices import *


tested_nx  = [5, 10, 20, 50]  # The tested number of x variables
tested_np  = [0, 5, 10]       # The tested number of p variables
tested_nw  = [5, 8]           # The tested number of w variables

@pytest.mark.parametrize("nx", tested_nx)
@pytest.mark.parametrize("np", tested_np)
@pytest.mark.parametrize("nw", tested_nw)
def testMasterVector(nx, np, nw):

    u = MasterVector(nx, np, nw)
    u.x = rng.rand(nx)
    u.p = rng.rand(np)
    u.w = rng.rand(nw)

    v = MasterVector(nx, np, nw)
    v.x = rng.rand(nx)
    v.p = rng.rand(np)
    v.w = rng.rand(nw)

    assert all(u.array() == npy.concatenate([u.x, u.p, u.w]))

    t = MasterVector(u); t += v
    assert t.array() == approx(u.array() + v.array())

    t = MasterVector(u); t -= v
    assert t.array() == approx(u.array() - v.array())

    t = MasterVector(u); t *= 1.234
    assert t.array() == approx(u.array() * 1.234)

    t = MasterVector(u); t /= 1.234
    assert t.array() == approx(u.array() / 1.234)

    t = u + v
    assert t.array() == approx(u.array() + v.array())

    t = u - v
    assert t.array() == approx(u.array() - v.array())

    t = u * 1.234
    assert t.array() == approx(u.array() * 1.234)

    t = 2.345 * u
    assert t.array() == approx(u.array() * 2.345)

    t = u / 1.234
    assert t.array() == approx(u.array() / 1.234)

    assert u.dot(v) == approx(sum(u.array() * v.array()))

    assert u.norm() == approx(npy.linalg.norm(u.array()))
    assert v.norm() == approx(npy.linalg.norm(v.array()))

    assert u.squaredNorm() == approx(u.norm() * u.norm())
    assert v.squaredNorm() == approx(v.norm() * v.norm())

    u.x[:nx] = 0.0  # ensure u.x[:nx] is a view to actual content, and not a copy
    u.p[:np] = 0.0  # ensure u.p[:np] is a view to actual content, and not a copy
    u.w[:nw] = 0.0  # ensure u.w[:nw] is a view to actual content, and not a copy

    assert all(u.x[:nx] == 0.0)
    assert all(u.p[:np] == 0.0)
    assert all(u.w[:nw] == 0.0)
